chimedb/dataset/get.py

```diff
@@ -21,42 +24,300 @@
+
+    _type = None
+    _base_dataset = None
+
+    @property
+    def type(self):
+        """Get the type of the attached dataset state."""
+        if self._type is None:
+            return self.dataset_state.type
+        return self._type
+
+    @property
+    def base_dataset(self):
+        """Get the base dataset."""
+        if not self._base_dataset:
+            self._base_dataset = self.base_dset
+        return self._base_dataset
+
+    @property
+    def dataset_state(self):
+        """Get the dataset state."""
+        # get the raw ID of the state (foreign key)
+        orm_base_class = super()
+        state_id = orm_base_class.state_id
+
+        if state_id not in _state_cache:
+            _state_cache[state_id] = orm_base_class.state
+        return _state_cache[state_id]
+
+    def __repr__(self):
+        if self._type:
+            return "<get.Dataset[{}]: {}>".format(self.type, self.id)
+        else:
+            return "<get.Dataset: {}>".format(self.id)
+
+    def closest_ancestor_of_type(self, type_):
+        """
+        Get the closest ancestorial dataset of the given type.
+
+        Parameters
+        ----------
+        type_ : str or :class:`chimedb.dataset.orm.DatasetStateType`
+
+        Raises
+        ------
+        :class:`chimedb.dataset.orm.DatasetStateType.DoesNotExist`
+            If `type_` is not an existing dataset state type.
+
+        Returns
+        -------
+        :class:`chimedb.dataset.orm.Dataset`
+            The closest ancestor of the given type. Returns `None` if no ancestor of the
+            given type found.
+        """
+        if isinstance(type_, str):
+            type_ = DatasetStateType.from_name(name=type_)
+            if type_ is None:
+                raise NotFoundError("{} is not a known DatasetStateType".format(type_))
+            type_ = type_.name
+        d = self
+
+        while d:
+            if d.type == type_:
+                return d
+            d = d.base_dataset
+
+        raise NotFoundError(
+            "No ancestor of type {} found for Dataset {}".format(type_, self.__repr__())
+        )
+
+
+def index():
+    """Pre-fetch and cache all Dataset(State(Type))s."""
+    query = (
+        orm.Dataset.select(orm.Dataset, orm.DatasetStateType.name.alias("_type"))
+        .join(orm.DatasetState)
+        .join(orm.DatasetStateType)
+    )
+
+    # Create a mapping from dataset ID to
+    dsdict = {d.id: d for d in query.objects(Dataset)}
+
+    # De-reference the base datasets
+    for d in dsdict.values():
+        base_dset = None if d.root or d.base_dset_id is None else dsdict[d.base_dset_id]
+        d._base_dataset = base_dset
+    _dataset_cache.update(dsdict)
+
+    # Pre-fetch all states without their .data and fill cache
+    for s in query.objects(DatasetState):
+        _state_cache[s.id] = s
+
+    # Pre-fetch and cache types
+    for t in query.objects(DatasetStateType):
+        _type_cache[t.id] = t
```

Comment:
I'm not quite sure how exactly this is meant to connect up, but these just essentially fill the state and type caches with the same content as the dataset cache.

It's performing exactly the same query each time, all the `.objects(DatasetState)` does is graft the attributes coming out of the query onto instances of the specified class (in this case DatasetState), but it's actually giving them the contents of a dataset state, e.g.
```
>>> list(dget._type_cache.values())[-1].base_dset
'9ce7e818a0cf17ba2794dbad2082dee0'
```

In this case you probably just want something like this (not tested):
```suggestion
    # Pre-fetch all states without their .data and fill cache
    for s in DatasetState.select(DatasetState.id, DatasetState.type):
        _state_cache[s.id] = s

    # Pre-fetch and cache types
    for t in DatasetStateType.select():
        _type_cache[t.id] = t
```